### PR TITLE
Web IDL: Check overloaded operations and redefined members

### PR DESCRIPTION
**Pull request by tidoust**
This completes the Web IDL study tool with checks to report on invalid overloaded operations and duplicated members within interface definitions but also across partials, mixins and partial mixins. Two new types of anomalies get reported: `overloaded` and `redefinedMember`.

As opposed to the test logic in Webref, no actual merge is being performed. Definitions are rather compared one by one, which makes it possible to report more fine-grained anomaly messages and in particular to target the spec that likely needs fixing.

Typically, when a partial dfn introduces a conflict with a base dfn, it seems logic to blame the partial dfn. Similarly, when a dfn introduces a conflict with one of its mixins, it seems logic to blame the dfn, since the dfn includes the mixin and not the other way round.

Only cases when multiple specs need to be blamed is when two specs create overloads in partial dfns (no a priori reason to consider that a partial has priority over another partial).

Note that the code will create multiple similar anomalies when the same method is overloaded in more than 2 partial dfns and in more than 2 specs. That is, theoretically, report could contain only one anomaly for the overloaded operation, linked to all specs that define the overload, but in practice report will contain one anomaly per pair of specs. That's not wrong per se, just not optimized. Such situations should not happen much in practice, if at all, unless the IDL starts being very clunky.

This update also rewrites tests to check on the number of anomalies first and report a dump of the report or anomaly when assertion fails. This is to ease debugging when a test fails.

Additional notes:
- <strike>I haven't tried the code on actual crawl results yet, so I wouldn't be surprised if I missed a few cases here and there. Hopefully, the logic is sound :)</strike> <strike>Code seems to yield consistent results on Webref non curated data, haven't tried yet with curated data.</strike> Code tested on curated Webref data. It reports 3 `incompatiblePartialIdlExposure` anomalies for MSE v2 (known spec issues, base interfaces need to be exposed in `DedicatedWorker`), which actually should have been detected by Webref's tests... It also reports a couple of `unexpectedEventHandler` anomalies (already reported as errors against the specs), and a number of enum anomalies (case or single value) that we will probably want to ignore in Webref.
- Checks on exposure to make sure that there is no overlap for members that have the same name but different exposure are still TBD. Same thing as in Webref tests!

### CODE REVIEW

File: test/study-webidl.js [hunk 1]
```diff
@@ -19,13 +20,32 @@ describe('The Web IDL analyser', async _ => {
...
+    }
+  }
 
   it('reports no anomaly if IDL is valid', async () => {
     const report = await analyzeIdl(`
 [Global=Window,Exposed=*]
 interface Valid {};
 `);
-    assert.deepEqual(report, []);
+    assertEqual(report, []);
```
Comment (dontcallmedom) [inline test/study-webidl.js:48]: This test passes "by mistake" - in `assertEqual`, both `actual` and `value` ends up being undefined, but no check on the actual content of `report` is done.

File: test/study-webidl.js [hunk 2]
```diff
@@ -34,13 +54,14 @@ interface Valid {};
 [Global=Window,Exposed=*]
 interface Invalid;
 `);
-    assert.deepEqual(report[0]?.name, 'invalid');
-    assert.deepEqual(report[0].message, `Syntax error at line 3, since \`interface Invalid\`:
+    assertEqual(report, 'length', 1);
```
Comment (dontcallmedom) [inline test/study-webidl.js:57]:
```suggestion
    assert.equal(report.length, 1);
```
I don't think assertEqual helps here

File: test/study-webidl.js [hunk 3]
```diff
@@ -56,8 +77,8 @@ interface Valid: Invalid {};
 interface Invalid{};
 `);
     const report = await studyWebIdl(crawlResult, curatedResult);
-    assert.deepEqual(report.length, 1);
-    assert.deepEqual(report[0]?.name, 'invalid');
+    assertEqual(report.length, 1);
```
Comment (dontcallmedom) [inline test/study-webidl.js:80]:
```suggestion
    assert.equal(report.length, 1);
```
I don't even want to think how this doesn't fail atm :)

File: test/study-webidl.js [hunk 4]
```diff
@@ -19,13 +20,32 @@ describe('The Web IDL analyser', async _ => {
     return studyWebIdl(crawlResult);
   }
 
+  function assertEqual(report, ...params) {
```
Comment (dontcallmedom) [inline test/study-webidl.js:23]: The name is confusing since it's not really checking equality in the general sense. Maybe `assertReportMatch`?

File: test/study-webidl.js (hunk 1 above)
Comment (dontcallmedom) [reply test/study-webidl.js:48]:
```suggestion
    assert.deepEqual(report, []);
```

File: test/study-webidl.js [hunk 5]
```diff
@@ -70,16 +91,17 @@ interface Invalid;
 [Global=Window,Exposed=*]
 interface Valid {};
 `);
-    assert.deepEqual(report, []);
+    assertEqual(report, []);
```
Comment (dontcallmedom) [inline test/study-webidl.js:94]:
```suggestion
    assert.deepEqual(report, []);
```
(see above)

File: test/study-webidl.js [hunk 6]
```diff
@@ -70,16 +91,17 @@ interface Invalid;
...
   });
 
 
   it('detects missing [Exposed] extended attributes', async () => {
     const report = await analyzeIdl(`
 interface Unexposed {};
 `);
-    assert.deepEqual(report[0]?.name, 'noExposure');
-    assert.deepEqual(report[0].message, 'The interface "Unexposed" has no [Exposed] extended attribute');
+    assertEqual(report, 'length', 1);
```
Comment (dontcallmedom) [inline test/study-webidl.js:102]:
```suggestion
    assert.equal(report.length, 1);
```

File: test/study-webidl.js [hunk 7]
```diff
@@ -88,8 +110,9 @@ interface Unexposed {};
 [Exposed=Unknown]
 interface WhereIAm {};
 `);
-    assert.deepEqual(report[0]?.name, 'unknownExposure');
-    assert.deepEqual(report[0].message, 'The [Exposed] extended attribute of the interface "WhereIAm" references unknown global(s): Unknown');
+    assertEqual(report, 'length', 1);
```
Comment (dontcallmedom) [inline test/study-webidl.js:113]:
```suggestion
    assert.equal(report.length, 1);
```

File: test/study-webidl.js [hunk 8]
```diff
@@ -107,19 +130,26 @@ interface Carlos : EventTarget {
 [Exposed=*]
 interface EventTarget {};
 `);
-    assert.deepEqual(report, []);
+    assertEqual(report, []);
```
Comment (dontcallmedom) [inline test/study-webidl.js:133]:
```suggestion
    assert.deepEqual(report, []);
```

File: test/study-webidl.js [hunk 9]
```diff
@@ -107,19 +130,26 @@ interface Carlos : EventTarget {
...
+typedef EventHandlerNonNull? EventHandler;
+
 [Global=Window,Exposed=*]
 interface Carlos {
   attribute EventHandler onbigbisous;
 };
 `);
-    assert.deepEqual(report[0]?.name, 'unexpectedEventHandler');
-    assert.deepEqual(report[0].message, 'The interface "Carlos" defines an event handler "onbigbisous" but does not inherit from EventTarget');
+    assertEqual(report, 'length', 1);
```
Comment (dontcallmedom) [inline test/study-webidl.js:150]:
```suggestion
    assert.equal(report.length, 1);
```

File: test/study-webidl.js [hunk 10]
```diff
@@ -142,9 +172,9 @@ partial interface Somewhere {};
 [Exposed=Elsewhere]
 partial interface MyPlace {};
 `);
-    assert.deepEqual(report[0]?.name, 'incompatiblePartialIdlExposure');
-    assert.deepEqual(report[0].message, 'The [Exposed] extended attribute of the partial interface "MyPlace" references globals on which the original interface is not exposed: Elsewhere (original exposure: Somewhere)');
-    assert.deepEqual(report.length, 1);
+    assertEqual(report.length, 1);
```
Comment (dontcallmedom) [inline test/study-webidl.js:175]:
```suggestion
    assert.equal(report.length, 1);
```

File: test/study-webidl.js [hunk 11]
```diff
@@ -158,9 +188,9 @@ interface Somewhere {};
 [Exposed=*]
 partial interface Somewhere {};
 `);
-    assert.deepEqual(report[0]?.name, 'incompatiblePartialIdlExposure');
-    assert.deepEqual(report[0].message, 'The partial interface "Somewhere" is exposed on all globals but the original interface is not (Somewhere)');
-    assert.deepEqual(report.length, 1);
+    assertEqual(report.length, 1);
```
Comment (dontcallmedom) [inline test/study-webidl.js:191]:
```suggestion
    assert.equal(report.length, 1);
```

File: test/study-webidl.js [hunk 12]
```diff
@@ -254,21 +284,21 @@ MyHome includes MyRoom;
...
   it('checks existence of target and mixin in includes statements', async () => {
     const report = await analyzeIdl(`
 MyHome includes MyRoom;
 `);
-    assert.deepEqual(report[0]?.name, 'unknownType');
-    assert.deepEqual(report[0].message, `Target "MyHome" in includes statement "MyHome includes MyRoom" is not defined anywhere`);
-    assert.deepEqual(report[1]?.name, 'unknownType');
-    assert.deepEqual(report[1].message, `Mixin "MyRoom" in includes statement "MyHome includes MyRoom" is not defined anywhere`);
-    assert.deepEqual(report.length, 2);
+    assertEqual(report.length, 2);
```
Comment (dontcallmedom) [inline test/study-webidl.js:297]:
```suggestion
    assert.equal(report.length, 2);
```

File: test/study-webidl.js [hunk 13]
```diff
@@ -278,11 +308,11 @@ dictionary MyHome { required boolean door; };
 
 MyHome includes MyRoom;
 `);
-    assert.deepEqual(report[0]?.name, 'wrongKind');
-    assert.deepEqual(report[0].message, `Target "MyHome" in includes statement "MyHome includes MyRoom" must be of kind "interface"`);
-    assert.deepEqual(report[1]?.name, 'wrongKind');
-    assert.deepEqual(report[1].message, `Mixin "MyRoom" in includes statement "MyHome includes MyRoom" must be of kind "interface mixin"`);
-    assert.deepEqual(report.length, 2);
+    assertEqual(report.length, 2);
```
Comment (dontcallmedom) [inline test/study-webidl.js:311]:
```suggestion
    assert.equal(report.length, 2);
```

File: test/study-webidl.js [hunk 14]
```diff
@@ -293,11 +323,11 @@ MyHome includes MyRoom;
 [Exposed=*] interface MyLivingRoom : MyRoom {};
 dictionary MyShelf : MyHome { required boolean full; };
 `);
-    assert.deepEqual(report[0]?.name, 'unknownType');
-    assert.deepEqual(report[0].message, `"MyLivingRoom" inherits from "MyRoom" which is not defined anywhere`);
-    assert.deepEqual(report[1]?.name, 'wrongKind');
-    assert.deepEqual(report[1].message, `"MyShelf" is of kind "dictionary" but inherits from "MyHome" which is of kind "interface"`);
-    assert.deepEqual(report.length, 2);
+    assertEqual(report.length, 2);
```
Comment (dontcallmedom) [inline test/study-webidl.js:326]:
```suggestion
    assert.equal(report.length, 2);
```

File: test/study-webidl.js [hunk 15]
```diff
@@ -311,15 +341,15 @@ dictionary MyShelf : MyHome { required boolean full; };
...
-    assert.deepEqual(report[0]?.name, 'unknownType');
-    assert.deepEqual(report[0].message, `Unknown type "bool" used in definition of "MyRoom"`);
-    assert.deepEqual(report[1]?.name, 'unknownType');
-    assert.deepEqual(report[1].message, `Unknown type "MyBed" used in definition of "MyRoom"`);
-    assert.deepEqual(report[2]?.name, 'unknownType');
-    assert.deepEqual(report[2].message, `Unknown type "MyUnknownType" used in definition of "MyRoom"`);
-    assert.deepEqual(report[3]?.name, 'unknownType');
-    assert.deepEqual(report[3].message, `Unknown type "UnknownInnerType" used in definition of "MyRoom"`);
-    assert.deepEqual(report.length, 4);
+    assertEqual(report.length, 4);
```
Comment (dontcallmedom) [inline test/study-webidl.js:344]:
```suggestion
    assert.equal(report.length, 4);
```

File: test/study-webidl.js [hunk 16]
```diff
@@ -336,15 +366,15 @@ interface mixin MyLivingRoom {};
...
-    assert.deepEqual(report.length, 4);
-    assert.deepEqual(report[0].name, 'redefinedWithDifferentTypes');
-    assert.deepEqual(report[0].message, `"MyNamespaceMixin" is defined multiple times with different types (namespace, interface mixin) in ${specUrl}`);
-    assert.deepEqual(report[1].name, 'wrongType');
-    assert.deepEqual(report[1].message, `Namespace "MyNamespace" cannot be used as a type in definition of "MyHome"`);
-    assert.deepEqual(report[2].name, 'wrongType');
-    assert.deepEqual(report[2].message, `Interface mixin "MyLivingRoom" cannot be used as a type in definition of "MyHome"`);
-    assert.deepEqual(report[3].name, 'wrongType');
-    assert.deepEqual(report[3].message, `Name "MyNamespaceMixin" exists but is not a type and cannot be used in definition of "MyHome"`);
+    assertEqual(report.length, 4);
```
Comment (dontcallmedom) [inline test/study-webidl.js:369]:
```suggestion
    assert.equal(report.length, 4);
```

File: test/study-webidl.js [hunk 17]
```diff
@@ -361,12 +391,318 @@ interface mixin MyNamespaceMixin {};
...
 };
 `);
-    assert.deepEqual(report[0]?.name, 'unknownExtAttr');
-    assert.deepEqual(report[0].message, `Unknown extended attribute "UnknownExtAttr" used in definition of "MyRoom"`);
-    assert.deepEqual(report[1]?.name, 'unknownExtAttr');
-    assert.deepEqual(report[1].message, `Unknown extended attribute "SuperUnknownExtAttr" used in definition of "MyLivingRoom"`);
-    assert.deepEqual(report[2]?.name, 'unknownExtAttr');
-    assert.deepEqual(report[2].message, `Unknown extended attribute "SuperUnknownExtAttr" used in definition of "MyBedRoom"`);
-    assert.deepEqual(report.length, 3);
+    assertEqual(report.length, 3);
```
Comment (dontcallmedom) [inline test/study-webidl.js:394]:
```suggestion
    assert.equal(report.length, 3);
```

File: test/study-webidl.js [hunk 18]
```diff
@@ -361,12 +391,318 @@ interface mixin MyNamespaceMixin {};
...
+
+partial interface MyPartialHome {
+  Promise<DOMString> overload();
+};
+
+partial interface MyPartialHome {
+  Promise<DOMString> overload(DOMString thing);
+};
+`);
+    assertEqual(report.length, 2);
```
Comment (dontcallmedom) [inline test/study-webidl.js:427]:
```suggestion
    assert.equal(report.length, 2);
```

File: test/study-webidl.js [hunk 19]
```diff
@@ -361,12 +391,318 @@ interface mixin MyNamespaceMixin {};
...
+  undefined overload(DOMString thing, boolean asap);
+};
+
+partial interface MyPartialHome {
+  Promise<DOMString> overload(DOMString thing);
+};
+`, specUrl2));
+    const report = await studyWebIdl(crawlResult);
+
+    assertEqual(report.length, 2);
```
Comment (dontcallmedom) [inline test/study-webidl.js:460]:
```suggestion
    assert.equal(report.length, 2);
```

File: test/study-webidl.js [hunk 20]
```diff
@@ -361,12 +391,318 @@ interface mixin MyNamespaceMixin {};
...
+
+interface mixin MyPartialRoom {
+  Promise<DOMString> overload();
+};
+
+partial interface mixin MyPartialRoom {
+  Promise<DOMString> overload(DOMString thing);
+};
+`);
+    assertEqual(report.length, 2);
```
Comment (dontcallmedom) [inline test/study-webidl.js:496]:
```suggestion
    assert.equal(report.length, 2);
```

File: test/study-webidl.js [hunk 21]
```diff
@@ -361,12 +391,318 @@ interface mixin MyNamespaceMixin {};
...
+  undefined overload(DOMString thing);
+};
+
+partial interface mixin MyPartialRoom {
+  Promise<DOMString> overload(DOMString thing, boolean asap);
+};
+`, specUrl2));
+    const report = await studyWebIdl(crawlResult);
+
+    assertEqual(report.length, 4);
```
Comment (dontcallmedom) [inline test/study-webidl.js:530]:
```suggestion
    assert.equal(report.length, 4);
```

File: test/study-webidl.js [hunk 22]
```diff
@@ -361,12 +391,318 @@ interface mixin MyNamespaceMixin {};
...
+
+interface mixin MyRoom {
+  undefined overload(DOMString thing);
+};
+
+partial interface mixin MyRoom {
+  undefined overload(DOMString thing, boolean asap);
+};
+`);
+    assertEqual(report.length, 6);
```
Comment (dontcallmedom) [inline test/study-webidl.js:574]:
```suggestion
    assert.equal(report.length, 6);
```

File: test/study-webidl.js [hunk 23]
```diff
@@ -361,12 +391,318 @@ interface mixin MyNamespaceMixin {};
...
+    assertEqual(report, 3, 'specs[0].url', specUrl2);
+    assertEqual(report, 3, 'specs[1].url', specUrl);
+  });
+
+
+  it('reports overloads across definitions (partial and mixin, same spec)', async () => {
+    const report = await analyzeIdl(`
+[Global=Home,Exposed=*]
+interface MyHome {
+  undefined overload();
```
Comment (dontcallmedom) [inline test/study-webidl.js:558]: the test could be made a bit shorter while still achieving its goal (IIUC) by not including `overload()`  in the main interface (and the main mixin)

File: test/study-webidl.js [hunk 24]
```diff
@@ -361,12 +391,318 @@ interface mixin MyNamespaceMixin {};
...
+  undefined overload(DOMString thing);
+};
+
+partial interface mixin MyRoom {
+  undefined overload(DOMString thing, boolean asap);
+};
+`, specUrl2));
+    const report = await studyWebIdl(crawlResult);
+
+    assertEqual(report.length, 6);
```
Comment (dontcallmedom) [inline test/study-webidl.js:612]:
```suggestion
    assert.equal(report.length, 6);
```

File: test/study-webidl.js [hunk 25]
```diff
@@ -361,12 +391,318 @@ interface mixin MyNamespaceMixin {};
...
+    assertEqual(report, 5, 'name', 'overloaded');
+    assertEqual(report, 5, 'message', '"operation overload" in partial interface mixin "MyRoom" overloads an operation defined in interface mixin "MyRoom"');
+  });
+
+
+  it('reports overloads across definitions (partial and mixin, different specs)', async () => {
+    const crawlResult = toCrawlResult(`
+[Global=Home,Exposed=*]
+interface MyHome {
+  undefined overload();
```
Comment (dontcallmedom) [inline test/study-webidl.js:594]: same as above re shorter test

File: test/study-webidl.js [hunk 26]
```diff
@@ -361,12 +391,318 @@ interface mixin MyNamespaceMixin {};
...
+    assertEqual(report, 5, 'specs.length', 1);
+    assertEqual(report, 5, 'specs[0].url', specUrl2);
+  });
+
+
+  it('reports redefined members (same spec)', async () => {
+    const report = await analyzeIdl(`
+[Global=Home,Exposed=*]
+interface MyHome {
+  undefined overload();
```
Comment (dontcallmedom) [inline test/study-webidl.js:649]: nit: maybe use a different name than `overload` (e.g. `redefined`)

File: test/study-webidl.js [hunk 27]
```diff
@@ -361,12 +391,318 @@ interface mixin MyNamespaceMixin {};
...
+
+partial interface MyHome {
+  attribute DOMString overload;
+};
+
+interface mixin MyRoom {
+  attribute unsigned long overload;
+};
+`);
+    assertEqual(report.length, 4);
```
Comment (dontcallmedom) [inline test/study-webidl.js:662]:
```suggestion
    assert.equal(report.length, 4);
```

File: test/study-webidl.js [hunk 28]
```diff
@@ -361,12 +391,318 @@ interface mixin MyNamespaceMixin {};
...
+  undefined overload(DOMString thing);
+};
+
+[Exposed=*]
+interface MyPartialHome {};
+
+partial interface MyPartialHome {
+  Promise<DOMString> overload();
+};
+`).concat(toCrawlResult(`
```
Comment (dontcallmedom) [inline test/study-webidl.js:449]: the `concat` pattern is used often enough that it may be better to redefine `toCrawlResult` to take two IDL declaration and assign the spec URL accordingly.

File: test/study-webidl.js [hunk 29]
```diff
@@ -361,12 +391,318 @@ interface mixin MyNamespaceMixin {};
...
+  attribute DOMString overload;
+};
+
+interface mixin MyRoom {
+  attribute unsigned long overload;
+};
+`, specUrl2));
+    const report = await studyWebIdl(crawlResult);
+
+    assertEqual(report, 'length', 3);
```
Comment (dontcallmedom) [inline test/study-webidl.js:692]:
```suggestion
    assert.equal(report.length, 3);
```

**Review by dontcallmedom — CHANGES_REQUESTED** [review]
Impressive! the test suite is in particular chief-kiss-worthy! a few bugs / clarifications worth fixing

File: src/lib/study-webidl.js [hunk 30]
```diff
@@ -285,6 +344,87 @@ async function studyWebIdl(edResults, curatedResults) {
...
+          // is supposed to pay attention to the other. We won't blame the
+          // "source" spec otherwise
+          if (target.idl.partial && source.idl.partial &&
+              target.idl.type === source.idl.type) {
+            targetName += ` (in ${specName(target.spec)})`;
+            specs.push(source.spec);
+          }
+        }
+
+        if (isOverloadedOperation(targetMember, sourceMember)) {
```
[the remark this answers is not in the document]
Comment (tidoust) [reply src/lib/study-webidl.js:393]: That was copied from tests in Webref but, looking into the Web IDL spec, I'm not even sure that the check on `Exposed` is actually correct. The Web IDL spec seems to forbid re-definitions of a member with the same identifier, regardless of their extended attributes, e.g. for [Constants](https://webidl.spec.whatwg.org/#idl-constants):

"The [identifier](https://webidl.spec.whatwg.org/#dfn-identifier) of a [constant](https://webidl.spec.whatwg.org/#dfn-constant) must not be the same as the identifier of another [interface member](https://webidl.spec.whatwg.org/#dfn-interface-member) or [callback interface member](https://webidl.spec.whatwg.org/#callback-interface-member) defined on the same [interface](https://webidl.spec.whatwg.org/#dfn-interface) or [callback interface](https://webidl.spec.whatwg.org/#dfn-callback-interface)"

In other words, I'd be tempted to drop the exposure check completely here.

(FWIW, the case does not arise in the Web IDL extracted in Webref).

File: src/lib/study-webidl.js [hunk 31]
```diff
@@ -285,6 +344,87 @@ async function studyWebIdl(edResults, curatedResults) {
...
+    const knownDuplicates = [];
+    if (!target.idl.members || !source.idl.members) {
+      return;
+    }
+    for (const targetMember of target.idl.members) {
+      if (!targetMember.name) {
+        continue;
+      }
+
+      for (const sourceMember of source.idl.members) {
```
[the remark this answers is not in the document]
Comment (tidoust) [reply src/lib/study-webidl.js:359]: I try to be consistent in the loops that are used: if outer loop is a `for`, I actually find it clearer not to mix with a `filter` and `forEach`. I suppose clarity is subjective ;) I'll rewrite.

I note I don't understand why this would be any faster: the call to `filter` would parse the entire list as well, same as the `for` loop. If anything, it will make things slower, since code will filter the list first, then go over the filtered list (`filter` followed by a `for`). But well, given the expected sizes of the lists at hand, any optimization would be overkill in any case.

[the remark this answers is not in the document]
Comment (dontcallmedom) [reply src/lib/study-webidl.js:359]: I don't mind keeping it as is, the consistency argument wfm

File: src/lib/study-webidl.js [hunk 32]
```diff
@@ -393,10 +533,25 @@ async function studyWebIdl(edResults, curatedResults) {
...
+      "callback interface",
+      "callback",
+      "dictionary",
+      "enum",
+      "interface",
+      "interface mixin",
+      "namespace",
+      "typedef"
+    ]);
+    if (!knownDfnKinds.has(mainDef.type)) {
```
[the remark this answers is not in the document]
Comment (tidoust) [reply src/lib/study-webidl.js:551]: I'll keep it, then ;) That supposes that a major version of WebIDL2 would be released, that we would actually notice that a new construct gets introduced, and that we would realize that we need to update the code accordingly (code won't fail until some actual Web IDL uses that new construct). I cannot think of a better place to protect us from ourselves...

File: test/study-webidl.js (hunk 2 above)
Comment (tidoust) [reply test/study-webidl.js:57]: It does because it will report the entire anomaly report to the console, which in turns allows to quickly understand why the assumption was incorrect. Without that, you would just get an "expected 1, got 2" message and you'd need to reproduce the run locally (if error occurred in CI) and switch to debug mode.

File: test/study-webidl.js (hunk 26 above)
Comment (tidoust) [reply test/study-webidl.js:649]: If I can't even create a bad WWII pun with a (misnamed) "operation overload"... :(
(Fun fact: [operation overload](https://en.wikipedia.org/wiki/Operation_Overload) seems to exist in practice, I didn't know ;))

File: test/study-webidl.js (hunk 2 above)
Comment (dontcallmedom) [reply test/study-webidl.js:57]: then I suggest an assertReportLength or something, rather than re-using the already somewhat convoluted assertEqual

File: test/study-webidl.js (hunk 26 above)
Comment (dontcallmedom) [reply test/study-webidl.js:649]: you can keep the pun for the overload test; you just have to find a different pun for redefined :)

File: src/lib/study-webidl.js (hunk 32 above)
[the remark this answers is not in the document]
Comment (dontcallmedom) [reply src/lib/study-webidl.js:551]: > I cannot think of a better place to protect us from ourselves

(clearly out of scope of this PR) - Most projects parsing WebIDL with webidl2js have the same need to detect this situation. I think the right way would be for Webidl2.js to expose its list of supported top-level types; or expose a walking tree API that could fail on un-matched types. But you're right that this wishful thinking as part of this PR :)

File: src/lib/study-webidl.js [hunk 33]
```diff
@@ -285,6 +344,87 @@ async function studyWebIdl(edResults, curatedResults) {
     }
   }
 
+  function checkMembers(target, source) {
+    const selfCheck = !source;
+    source = source ?? target;
```
[the remark this answers is not in the document]
Comment (dontcallmedom) [reply src/lib/study-webidl.js:349]: or even
```js
function checkMembers(target, source = target) {
  const selfCheck = source === target;
```

[the remark this answers is not in the document]
Comment (tidoust) [reply src/lib/study-webidl.js:347]: I used the first approach. Second approach would not catch calls such as `checkMembers(obj, null)` as [documented on MDN](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Functions/Default_parameters#passing_undefined_vs._other_falsy_values), which I always find confusing.

File: test/study-webidl.js (hunk 1 above)
Comment (tidoust) [reply test/study-webidl.js]: Now using a more explicit `assertNbAnomalies(report, 0)`

File: test/study-webidl.js (hunk 4 above)
Comment (tidoust) [reply test/study-webidl.js]: I switched to another approach with an `assertAnomaly` function that compares objects directly instead of the more atomic approach I had.

File: test/study-webidl.js (hunk 26 above)
Comment (tidoust) [reply test/study-webidl.js]: Ah, I hadn't spotted the difference. Good, then `dejaVu` it is :)
